src/kudu/consensus/log_cache.cc

```diff
@@ -315,7 +315,16 @@ Status LogCache::ReadOps(int64_t after_op_index,
                          std::vector<ReplicateRefPtr>* messages,
                          OpId* preceding_op) {
   DCHECK_GE(after_op_index, 0);
-  RETURN_NOT_OK(LookupOpId(after_op_index, preceding_op));
+
```

Comment:
more for my understanding: When is ReadOps used? Is it when we have already missed the entries in the LogCache?

Reply:
No. LogCache::ReadOps() is called whenever ops need to be fetched from log/log-cache for shipping. On cache miss, the method internally calls into log::ReadReplicatesInRange() to fetch the ops